{"cells":[{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["####################### VARIANTS PART #####################################################"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":["import pandas as pd\n","import glob\n","from scripts.explode import explode\n","import sys\n","import os\n","import re\n","\n","import pandas as pd\n","import numpy as np\n","\n","\n"]},{"cell_type":"code","execution_count":66,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"10_123810032_C/T T ENSG00000138162 ENST00000369005 Transcript missense_variant 453 113 38 T/M aCg/aTg rs540521848,COSM465374,COSM2148937\n10_133967449_C/T T ENSG00000188385 ENST00000298622 Transcript synonymous_variant 2307 2169 723 D gaC/gaT rs763531628,COSM2148890\n11_124489539_G/A A ENSG00000154143 ENST00000284288 Transcript missense_variant 954 887 296 R/Q cGg/cAg COSM211013\n11_47380512_G/T T ENSG00000066336 ENST00000227163 Transcript missense_variant 417 379 127 P/T Cca/Aca .\n11_89868837_C/T T ENSG00000077616 ENST00000534061 Transcript missense_variant 423 193 65 R/C Cgt/Tgt COSM932955\n11_92570936_G/A A ENSG00000165323 ENST00000298047 Transcript synonymous_variant 10349 10332 3444 P ccG/ccA rs374103134,COSM2148917,COSM2148916,COSM2148915\n12_107371855_A/G G ENSG00000120832 ENST00000552029 Transcript missense_variant 2707 638 213 L/S tTa/tCa COSM2148901\n12_108012011_G/A A ENSG00000151136 ENST00000280758 Transcript missense_variant 2836 2308 770 E/K Gag/Aag COSM2148894\n12_117768962_C/A A ENSG00000089250 ENST00000338101 Transcript upstream_gene_variant . . . . . .\n"}],"source":["mafFields = ['Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build',\n","             'Chromosome', 'Start_position', 'End_position', 'Strand',\n","             'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n","             'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS', 'dbSNP_Val_Status',\n","             'Tumor_Sample_Barcode', 'Normal_Norm_SAmple_Barcode',\n","             'Match_Norm_Seq_Allele1', 'Match_Norm_Seq_Allele2', 'Tumor_Validation_Allele1',\n","             'Tumor_Validation_Allele2', 'Match_Norm_Validation_Allele1', 'Match_Norm_Validation_Allele2',\n","             'Verification_Status', 'Validation_Status', 'Mutation_Status', 'Sequencing_Phase', 'Sequence_Source',\n","             'Validation_Method', 'Score', 'BAM_File', 'Sequencer', 'Tumor_Sample_UUID', 'Matched_Norm_Sample_UUID']\n","\n","maf = \"/home/vruizser/head_mc3.maf\"\n","\n","up_cols=[\"Chromosome\", \"Start_Position\", \"Reference_Allele\", \"Allele\"]\n","loc_cols = [\"Chromosome\", \"Start_Position\", \"End_Position\"]\n","rest_of_cols = [\"Allele\",\n","                \"Gene\",\n","                \"Feature\",\n","                \"Feature_type\",\n","                \"Consequence\",\n","                \"cDNA_position\",\n","                \"CDS_position\",\n","                \"Protein_position\",\n","                \"Amino_acids\",\n","                \"Codons\",\n","                \"Existing_variation\"]\n","\n","with open(maf) as f:\n","    # get col names\n","    cols = f.readline().split('\\t')\n","    uploaded_variation_cols=np.where(np.isin(cols,up_cols))\n","    location_cols = np.where(np.isin(cols,up_cols))\n","    rest_cols=np.where(np.isin(cols, rest_of_cols))\n","    # parse file\n","    for line in f:\n","        #print(line.split('\\t')[N[0][0]] + '_' + line.split('\\t')[N[0][0]])\n","        all_s = np.asarray(line.split('\\t'))[uploaded_variation_cols]\n","        rest_s = np.asarray(line.split('\\t'))[rest_cols]\n","        print( '_'.join(all_s[:-1]) + '/' + all_s[-1], ' '.join(rest_s))\n","        \n","        #print(pd.DataFrame(list(line)))\n","\n","\n","#'awk 'NR==1 { \\\n","#    for (i=1; i<=NF; i++) { \\\n","#        f[$i] = i \\\n","#    } \\\n","#    } \\\n","#{ print $(f[\"Consequence\"])\"_\"$(f[\"Start_Position\"]) } \\\n","#' head_mc3.maf '\n","\n","#detect_column = \"awk -F ' ' '{{for(i=1;i<=NF;i++) {{if ($i ~ /{}/){{print i; exit}}}}}}' {}\"\n","\n","#print(detect_colum)\n"]}],"nbformat":4,"nbformat_minor":2,"metadata":{"language_info":{"name":"python","codemirror_mode":{"name":"ipython","version":3}},"orig_nbformat":2,"file_extension":".py","mimetype":"text/x-python","name":"python","npconvert_exporter":"python","pygments_lexer":"ipython3","version":3}}