{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"import mapping_tools as mt\nimport numpy as np\nimport sys\nimport os\nimport gzip\nimport re\nimport pandas as pd\nfrom timeit import default_timer as timer\nfrom VEPcrossref import VEPfileCrossrefGenerator as cr"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# INTERFACES PART"},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":"# Extract the info corresponding to the prot ID (Interface parse)\n\n#sys.argv[1]\n#sys.argv[2]\n\nprotID = 'ENSP00000482258'\nint_db_file = '/home/vruizser/PhD/2018-2019/Immunity_interfaces_analysis/raw_data/interfaces_mapped_to_v94.csv'\n\n  \ninterfacesDB = open(int_db_file, \"r\")\ncols = interfacesDB.readline().strip().split(\" \")\n#prot_interface = mt.ProtIntDB_parser(interfacesDB, protID)\nprot_interface = mt.parser(interfacesDB,  protID, cols, \" \" )"},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[],"source":"subset_prot_interface = prot_interface[['pdb.id',\n                      'ensembl.prot.id',\n                      'temp.chain',\n                      'int.chain',\n                      'interaction',\n                      'resid_sseq',\n                      'mapped.real.pos',\n                      'pdb.pos' ]]\n\n# put it into right format\nsubset_prot_interface.columns = subset_prot_interface.columns.str.replace(\"\\\\.\", \"_\")\n\nfor col in (\"resid_sseq\", \"mapped_real_pos\", \"pdb_pos\"):\n    subset_prot_interface.loc[ : , col] = subset_prot_interface[col].str.replace(\"-\", ',')\n    subset_prot_interface.loc[ : , col] = subset_prot_interface[col].str.split(',')\n\nsubset_prot_interface = mt.explode(subset_prot_interface, [\"resid_sseq\", \"mapped_real_pos\", \"pdb_pos\"])\n\nsubset_prot_interface.rename(columns={'mapped_real_pos':'Protein_position'}, inplace=True)\n\nsubset_prot_interface['region_id'] = subset_prot_interface['pdb_id'] + '_' + subset_prot_interface['ensembl_prot_id'] + '_' + subset_prot_interface['temp_chain'] + '_' + subset_prot_interface['int_chain'] + '_' + subset_prot_interface['interaction']"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"####################### VARIANTS PART #####################################################"},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[],"source":"# Translate protID to gene ID\n#subset_prot_interface"},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[],"source":"biomartdb = open(\"/home/vruizser/PhD/2018-2019/git/PDBmapper/project/gene_transcript_protein_ens_ids.txt\", \"r\")\nensemblIDs = mt.ensemblID_translator(biomartdb, protID)"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[],"source":"# get the corresponding VEP file\ncrossref_file = open(\"/home/vruizser/PhD/2018-2019/git/PDBmapper/project/geneids_VEPfiles_crossref.txt\", \"r\")\nVEP_filename = mt.VEP_getter(crossref_file, ensemblIDs[\"geneID\"])"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":"VEP_dir= \"/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/\"\nVEPfile = open(VEP_dir + VEP_filename, \"r\")\ngeneID = ensemblIDs['geneID']\ncols = pd.read_csv(VEPfile, nrows = 0, skiprows = 42, sep = \"\\t\").columns\nVEP = mt.parser(VEPfile,  geneID, cols, \"\\t\" )"},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":"setID = MapVariantToPDB(VEP, subset_prot_interface, 'region_id')"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# define main function to execute the previous defined functions together\ndef main():\n\n \n\n    \n    \n    return\n\n\n##########################\n# execute main function  #\n##########################\nif __name__ == '__main__':\n    main()"},{"cell_type":"code","execution_count":45,"metadata":{},"outputs":[],"source":"import vcfpy\nimport os\nimport re"},{"cell_type":"code","execution_count":130,"metadata":{},"outputs":[],"source":"input_vcf = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\n#input_vcf = \"/home/vruizser/PhD/2019-2020/varMapDB/clinvar_20190828.vcf.gz\"\n\n#line = '#Uploaded_variation\tLocation\tAllele\tGene\tFeature\tFeature_type\tConsequence\tcDNA_position\tCDS_position\tProtein_position\tAmino_acids\tCodons\tExisting_variation\tExtra'\n#re.search(vep_header, line)\n#print(  vep_header in line. split())\n#result =  all(elem in vep_header  for elem in line.split())\n\n#print(result)\ndef recognize_format(infile):\n    try:\n        vcf_reader = vcfpy.Reader.from_path(infile)\n    \n        if 'CSQ' not in vcf_reader.header.info_ids():\n        #    vcf_reader.close()\n            return print(\"ERROR: VCF {} is not VEP-annotated. Please annotate the VCF with VEP before running this tool.\".format(infile))\n        else : \n            return print('is VCF')\n    except:\n        print('Not in vcf format. Checking if VEP file')\n        vep_header = set(['#Uploaded_variation',\n                                'Location',\n                                'Allele',\n                                'Gene',\n                                'Feature',\n                                'Feature_type',\n                                'Consequence',\n                                'cDNA_position',\n                                'CDS_position',\n                                'Protein_position',\n                                'Amino_acids',\n                                'Codons',\n                                'Existing_variation'])\n            \n        input_vcf= open(infile,'r')\n        for line in input_vcf:\n            jeje = set(line.split())\n                \n            res = vep_header.issubset(jeje)\n            if res:\n                return print(\"is vep format\") \n\n\n        return print(\"input format not recognized\")\n\n\nrecognize_format(input_vcf)\n\n#except:\n#    print('Incorrect input file format. The possible inputs are either VCF or VEP files')\n#    exit(-1)"},{"cell_type":"code","execution_count":139,"metadata":{},"outputs":[],"source":"from scripts.detect_vcf_format import detect_format\nimport vcfpy\nimport subprocess\ninput_vcf = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\n\nprint('Detecting inputs variants file format...')\ninput_format = detect_format(input_vcf)\n"},{"cell_type":"code","execution_count":27,"metadata":{},"outputs":[],"source":"from subprocess import call\nfrom scripts.detect_vcf_format import detect_format\nimport os\nimport sys"},{"cell_type":"code","execution_count":125,"metadata":{},"outputs":[],"source":"input_file = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\n\nfrom scripts.detect_vcf_format import detect_format\nimport subprocess\n\n#print('Detecting inputs variants file format...')\n#input_format = detect_format(input_file)\n#if input_format == \"vcf\":\n#    print(input_format)\n    \n        #call('sh', './vcf_to_vep.sh', input_file)\n        #call('sh', 'split_vep_by_protid.sh', input_file)\n#elif input_format == \"vep\":\n#print(input_format)\n    #call(['./split_vep_by_protid.sh', input_file] )\n    #os.system(\"sh ./split_vep_by_protid.sh titiri\")\nprint(input_file) \nCOMMAND='sh ./split_vep_by_protid.sh {}'.format(input_file)\n    #subprocess.run([\"ls\", \"-l\"])\n    #CompletedProcess(args=['ls', '-l'], returncode=0)\n#logfile = open('log.txt', 'w')\np = subprocess.Popen(COMMAND, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)\n#ret_code = p.wait()\n#logfile.flush()\nOUT, ERR = p.communicate()\n\nprint(OUT)\nprint(ERR)\n\n#import os\n#cmd = 'wc -l my_text_file.txt > out_file.txt'\n#os.system(COMMAND)\n\n#print(p.wait())\n#OUT, ERR = p.communicate()\n    #OUT, ERR = p.communicate()\n#print(OUT.decode('utf8'))\n\n\n#else: \n#    print(input_format)\n\n"},{"cell_type":"code","execution_count":44,"metadata":{},"outputs":[],"source":"import subprocess\nimport os\n\ninput_file = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\nout_dir='./scripts/input_pdbmapper/splitted_vcf_db/'\nos.makedirs(out_dir, exist_ok=True)\n\n\ncommand = \"awk -F ' ' '{{for(i=1;i<=NF;i++) {{if ($i ~ /ENSG/){{print i; exit}}}}}}' {} \".format(input_file)\n\n\nprint(command)\n\n# First command\np1 = subprocess.Popen(command, stdout=subprocess.PIPE, shell = True)\n# Second command's input linked to the first one's output\n\n# Read from p2 to get the output\nout, err = p1.communicate()\n#print(p1.stdout)\ncommand2 = \"grep -v '##' {} | \\\nawk -v ci=\\\"{}\\\" \\\n-v od=\\\"{}\\\" \\\n-F ' ' 'NR==1 \\\n{{h=$0; next}} \\\n{{f=od$ci\\\".vep\\\"}} !($ci in p) \\\n{{p[$ci]; print h > f}} \\\n{{print >> f; close(f)}}'\".format(input_file, out.decode('utf8'), out_dir) \n\n#command2 = \"grep -v '##' {}\".format(input_file) \n\nprint(command2)\np2 = subprocess.Popen(command2, stdout=subprocess.PIPE, shell = True)\n\n# Read from p2 to get the output\nout2, err2 = p2.communicate()\n\n\n"},{"cell_type":"code","execution_count":203,"metadata":{},"outputs":[],"source":"input_file = '/home/vruizser/PhD/2019-2020/ExAC_benchmark/ExAC.r1.sites.vep.vcf'\nout_dir='./scripts/input_pdbmapper/'\nout_file = out_dir + 'converted_vcf.vep'\n\ncommand = \"export BCFTOOLS_PLUGINS=./required_packages/bcftools/plugins\"\n#print(command + '\\n')\n\n# First command\n\n#p1 = subprocess.Popen(command ).wait()\n\nos.environ[\"BCFTOOLS_PLUGINS\"] = \"./required_packages/bcftools/plugins\"\n\ncommand2 = \"./required_packages/bcftools/bcftools +split-vep {0} \\\n-o {1} \\\n-f '%{2}\\_%{3}\\_%{4}\\/%{5} \\\n%{2}:%{3} %{6} %{7} %{8} %{9} %{10} %{11} \\\n%{12} %{13} %{14} %{15} %{16}\\\\n' \\\n-A tab -d\" .format(input_file, out_file, \"CHROM\", \"POS\", \"REF\", \"ALT\",\n                   \"Allele\", \"Gene\", \"Feature\", \"Feature_type\", \"Consequence\",\n                   \"cDNA_position\", \"CDS_position\", \"Protein_position\",\n                    \"Amino_acids\", \"Codons\", \"Existing_variation\" )\n\n#command2= \"./required_packages/bcftools/bcftools filter -H -r '%s':'%s'-'%s' '%s' -o '%s'\" % \"CHROM\", \"START\", \"END\", input_file, out_file\ncommand2= \"./required_packages/bcftools/bcftools \\\n+split-vep /home/vruizser/PhD/2019-2020/ExAC_benchmark/ExAC.r1.sites.vep.vcf \\\n-o ./scripts/input_pdbmapper/converted_vcf.vep \\\n-f '%{}\\_%{}\\_%REF\\/%ALT %Allele %Gene\\\\n' -A tab -d\" .format(\"CHROM\", \"POS\")\n\n#%Allele %Gene %Feature %Feature_type %Consequence %cDNA_position \\\n#%CDS_position %Protein_position %Amino_acids %Codons %Existing_variation\n#print(command2)\nlog = open('./scripts/input_pdbmapper/some file.txt', 'a')\nlog.write('some text, as header of the file\\n')\nlog.flush() \n\n\n\n#out1, err1 =p1.communicate()\n\np2 = subprocess.Popen(command2, stdout=subprocess.PIPE,stderr=log, shell = True)\n\nout2, err2 = p2.communicate()\n\n#f = open(out_file, \"w\")\n#f.write(out2)\n#f.close()\n\nprint(out2)\nprint(command2)\n\n#print(\"/home/vruizser/PhD/2018-2019/git/PDBmapper/required_packages/bcftools/bcftools +split-vep {}\".format(input_file))\n#print(command2)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"export BCFTOOLS_PLUGINS=./required_packages/bcftools/plugins \n\n./required_packages/bcftools/bcftools +split-vep $VCF_FILE \\\n                                       -o $OUT_FILENAME \\\n                                       -f '%CHROM\\_%POS\\_%REF\\/%ALT \\\n%CHROM:%POS %Allele %%Gene %Feature %Feature_type %Consequence %cDNA_position \\\n%CDS_position %Protein_position %Amino_acids %Codons %Existing_variation\\n'\n-A tab -d\n\n# add header\ngawk -i inplace '{if(NR==1){$0=\"#Uploaded_variation Location Allele Gene \\\nFeature Feature_type Consequence cDNA_position CDS_position Protein_position \\\nAmino_acids Codons Existing_variation\\n\"$0; print $0};\n                  if(NR!=1){print $0}}' $OUT_FILENAME\n\n\ncommand2 = \"./required_packages/bcftools/bcftools +split-vep {} \\\n-o {} \\\n-f '%s\\_%s\\_%s\\/%s %s:%s %s %s %s %s %s %s %s %s %s %s' \\\n-A tab -d\" % ('%CHROM', 'POS', 'REF', 'ALT', 'CHROM', 'POS', 'Allele', 'Gene', 'Feature', 'Feature_type', 'Consequence', 'cDNA_position', \n'CDS_position', 'Protein_position', 'Amino_acids' ,'Codons' ,'Existing_variation', 'input_file', 'out_file')"},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"input_file = '/home/vruizser/PhD/2019-2020/ExAC_benchmark/ExAC.r1.sites.vep.vcf'\ninput_file = '/home/vruizser/PhD/2019-2020/ExAC_benchmark/head_vcf.txt'\nout_dir='./scripts/input_pdbmapper/'\nout_file = out_dir + 'converted_vcf.vep2'\n\nfrom scripts.vcf_to_vep import vcf_to_vep\n\nvcf_to_vep(input_file, out_dir, out_file)"}],"nbformat":4,"nbformat_minor":2,"metadata":{"language_info":{"name":"python","codemirror_mode":{"name":"ipython","version":3}},"orig_nbformat":2,"file_extension":".py","mimetype":"text/x-python","name":"python","npconvert_exporter":"python","pygments_lexer":"ipython3","version":3}}