{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"import mapping_tools as mt\nimport numpy as np\nimport sys\nimport os\nimport gzip\nimport re\nimport pandas as pd\nfrom timeit import default_timer as timer\nfrom VEPcrossref import VEPfileCrossrefGenerator as cr"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# INTERFACES PART"},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":"# Extract the info corresponding to the prot ID (Interface parse)\n\n#sys.argv[1]\n#sys.argv[2]\n\nprotID = 'ENSP00000482258'\nint_db_file = '/home/vruizser/PhD/2018-2019/Immunity_interfaces_analysis/raw_data/interfaces_mapped_to_v94.csv'\n\n  \ninterfacesDB = open(int_db_file, \"r\")\ncols = interfacesDB.readline().strip().split(\" \")\n#prot_interface = mt.ProtIntDB_parser(interfacesDB, protID)\nprot_interface = mt.parser(interfacesDB,  protID, cols, \" \" )"},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[],"source":"subset_prot_interface = prot_interface[['pdb.id',\n                      'ensembl.prot.id',\n                      'temp.chain',\n                      'int.chain',\n                      'interaction',\n                      'resid_sseq',\n                      'mapped.real.pos',\n                      'pdb.pos' ]]\n\n# put it into right format\nsubset_prot_interface.columns = subset_prot_interface.columns.str.replace(\"\\\\.\", \"_\")\n\nfor col in (\"resid_sseq\", \"mapped_real_pos\", \"pdb_pos\"):\n    subset_prot_interface.loc[ : , col] = subset_prot_interface[col].str.replace(\"-\", ',')\n    subset_prot_interface.loc[ : , col] = subset_prot_interface[col].str.split(',')\n\nsubset_prot_interface = mt.explode(subset_prot_interface, [\"resid_sseq\", \"mapped_real_pos\", \"pdb_pos\"])\n\nsubset_prot_interface.rename(columns={'mapped_real_pos':'Protein_position'}, inplace=True)\n\nsubset_prot_interface['region_id'] = subset_prot_interface['pdb_id'] + '_' + subset_prot_interface['ensembl_prot_id'] + '_' + subset_prot_interface['temp_chain'] + '_' + subset_prot_interface['int_chain'] + '_' + subset_prot_interface['interaction']"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"####################### VARIANTS PART #####################################################"},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[],"source":"# Translate protID to gene ID\n#subset_prot_interface"},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[],"source":"biomartdb = open(\"/home/vruizser/PhD/2018-2019/git/PDBmapper/project/gene_transcript_protein_ens_ids.txt\", \"r\")\nensemblIDs = mt.ensemblID_translator(biomartdb, protID)"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[],"source":"# get the corresponding VEP file\ncrossref_file = open(\"/home/vruizser/PhD/2018-2019/git/PDBmapper/project/geneids_VEPfiles_crossref.txt\", \"r\")\nVEP_filename = mt.VEP_getter(crossref_file, ensemblIDs[\"geneID\"])"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":"VEP_dir= \"/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/\"\nVEPfile = open(VEP_dir + VEP_filename, \"r\")\ngeneID = ensemblIDs['geneID']\ncols = pd.read_csv(VEPfile, nrows = 0, skiprows = 42, sep = \"\\t\").columns\nVEP = mt.parser(VEPfile,  geneID, cols, \"\\t\" )"},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":"setID = MapVariantToPDB(VEP, subset_prot_interface, 'region_id')"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# define main function to execute the previous defined functions together\ndef main():\n\n \n\n    \n    \n    return\n\n\n##########################\n# execute main function  #\n##########################\nif __name__ == '__main__':\n    main()"},{"cell_type":"code","execution_count":45,"metadata":{},"outputs":[],"source":"import vcfpy\nimport os\nimport re"},{"cell_type":"code","execution_count":130,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"Not in vcf format. Checking if VEP file\nis vep format\n"}],"source":"input_vcf = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\n#input_vcf = \"/home/vruizser/PhD/2019-2020/varMapDB/clinvar_20190828.vcf.gz\"\n\n#line = '#Uploaded_variation\tLocation\tAllele\tGene\tFeature\tFeature_type\tConsequence\tcDNA_position\tCDS_position\tProtein_position\tAmino_acids\tCodons\tExisting_variation\tExtra'\n#re.search(vep_header, line)\n#print(  vep_header in line. split())\n#result =  all(elem in vep_header  for elem in line.split())\n\n#print(result)\ndef recognize_format(infile):\n    try:\n        vcf_reader = vcfpy.Reader.from_path(infile)\n    \n        if 'CSQ' not in vcf_reader.header.info_ids():\n        #    vcf_reader.close()\n            return print(\"ERROR: VCF {} is not VEP-annotated. Please annotate the VCF with VEP before running this tool.\".format(infile))\n        else : \n            return print('is VCF')\n    except:\n        print('Not in vcf format. Checking if VEP file')\n        vep_header = set(['#Uploaded_variation',\n                                'Location',\n                                'Allele',\n                                'Gene',\n                                'Feature',\n                                'Feature_type',\n                                'Consequence',\n                                'cDNA_position',\n                                'CDS_position',\n                                'Protein_position',\n                                'Amino_acids',\n                                'Codons',\n                                'Existing_variation'])\n            \n        input_vcf= open(infile,'r')\n        for line in input_vcf:\n            jeje = set(line.split())\n                \n            res = vep_header.issubset(jeje)\n            if res:\n                return print(\"is vep format\") \n\n\n        return print(\"input format not recognized\")\n\n\nrecognize_format(input_vcf)\n\n#except:\n#    print('Incorrect input file format. The possible inputs are either VCF or VEP files')\n#    exit(-1)"},{"cell_type":"code","execution_count":139,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"Detecting inputs variants file format...\nThe input file is not in vcf format.               Checking if it is a VEP file...\nis vep format\n"}],"source":"from scripts.detect_vcf_format import detect_format\nimport vcfpy\nimport subprocess\ninput_vcf = '/home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'\n\nprint('Detecting inputs variants file format...')\ninput_format = detect_format(input_vcf)\n"},{"cell_type":"code","execution_count":145,"metadata":{},"outputs":[{"ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: './home/vruizser/PhD/2018-2019/git/PDBmapper/scripts/split_vep_by_protid.sh /home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000': './home/vruizser/PhD/2018-2019/git/PDBmapper/scripts/split_vep_by_protid.sh /home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-145-26f8a728bcbb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;31m# no block, it start a sub process.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;31m# and you can block util the cmd execute finish\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36mcall\u001b[0;34m(timeout, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    285\u001b[0m     \u001b[0mretcode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ls\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"-l\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    286\u001b[0m     \"\"\"\n\u001b[0;32m--> 287\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    288\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors)\u001b[0m\n\u001b[1;32m    727\u001b[0m                                 \u001b[0mc2pread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2pwrite\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    728\u001b[0m                                 \u001b[0merrread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrwrite\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 729\u001b[0;31m                                 restore_signals, start_new_session)\n\u001b[0m\u001b[1;32m    730\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    731\u001b[0m             \u001b[0;31m# Cleanup if the child failed starting.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36m_execute_child\u001b[0;34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)\u001b[0m\n\u001b[1;32m   1362\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0merrno_num\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mENOENT\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1363\u001b[0m                             \u001b[0merr_msg\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m': '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1364\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno_num\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1365\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1366\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './home/vruizser/PhD/2018-2019/git/PDBmapper/scripts/split_vep_by_protid.sh /home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000': './home/vruizser/PhD/2018-2019/git/PDBmapper/scripts/split_vep_by_protid.sh /home/vruizser/PhD/2018-2019/PanCancer_data/vep_output/vep_PanCan_chr_1_1-100000'"]}],"source":"if input_format == \"vcf\":\n    call('sh', './vcf_to_vep.sh', input_vcf )\n    call('sh', 'split_vep_by_protid.sh', './data/converted_vcf.vep' )\nelif input_format == \"vep\":\n    cmd = \"./home/vruizser/PhD/2018-2019/git/PDBmapper/scripts/split_vep_by_protid.sh \" + input_vcf\n\n    # no block, it start a sub process.\n    p = subprocess.call(cmd)\n\n    # and you can block util the cmd execute finish\n    p\n\nelse: \n    print(input_format)"}],"nbformat":4,"nbformat_minor":2,"metadata":{"language_info":{"name":"python","codemirror_mode":{"name":"ipython","version":3}},"orig_nbformat":2,"file_extension":".py","mimetype":"text/x-python","name":"python","npconvert_exporter":"python","pygments_lexer":"ipython3","version":3}}